src/analyzer.ts

```diff
@@ -1063,21 +1064,36 @@ export class Analyzer {
           dependencyStillUnvalidated = dependency.config;
           continue;
         }
-        const result = this._checkForCyclesAndSortDependencies(
-          dependency.config,
-          trail,
-          // Walk through no-command scripts when determining if something is
-          // being directly invoked (e.g. if the top-level script has no command
-          // and simply delegates to one or more other scripts, then those
-          // dependencies are effectively being directly invoked).
-          isDirectlyInvoked && config.command === undefined
-        );
-        if (!result.ok) {
+        const validDependencyConfigResult =
+          this._checkForCyclesAndSortDependencies(
+            dependency.config,
+            trail,
+            // Walk through no-command scripts when determining if something is
+            // being directly invoked (e.g. if the top-level script has no command
+            // and simply delegates to one or more other scripts, then those
+            // dependencies are effectively being directly invoked).
+            isDirectlyInvoked && config.command === undefined
+          );
+        if (!validDependencyConfigResult.ok) {
           return {
             ok: false,
-            error: this._markAsInvalid(config, result.error.dependencyFailure),
+            error: this._markAsInvalid(
+              config,
+              validDependencyConfigResult.error.dependencyFailure
+            ),
           };
         }
+        const validDependencyConfig = validDependencyConfigResult.value;
+        if (validDependencyConfig.service) {
+          // We directly depend on a service.
+          config.effectiveServiceDependencies.push(validDependencyConfig);
+        } else if (validDependencyConfig.command === undefined) {
```

Comment:
So effectiveServiceDependencies do not include transitive service dependencies, except in the no-command case?

Reply:
Yeah, exactly.

We traverse through no-command scripts when doing this analysis. This is true pretty generally when thinking about the build graph, because a no-command script is basically just a shorthand alias for a collection of other scripts.